### PR TITLE
Optional attendees code

### PR DESCRIPTION
**Pull request by meghakak**
This branch adds the optional attendees feature and passes all previously added tests.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -35,22 +35,42 @@
...
  
     // No unavailable time ranges means the entire day is available
-    if (unavailableTimes.isEmpty()) {
+    if (unavailableTimes.isEmpty() && optionalUnavailableTimes.isEmpty()) {
       return ImmutableList.of(TimeRange.WHOLE_DAY);
     }
 
-    // TODO: Change structure to not update a class variable in the stream
     // Find all available time ranges for the requested meeting based on the attendees' unavailable times 
+    ImmutableList<TimeRange> availableTimes = ImmutableList.copyOf(getAvailableTimeRanges(unavailableTimes, request));
```
Comment (linhnguye-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It looks like you always convert the result of getAvailableTimeRanges to an ImmutableList?

If you do not expect to ever need to change the result of getAvailableTimeRanges I would change the method to return an ImmutableList<TimeRange> instead

**Review by linhnguye-g — APPROVED** [review]
LGTM

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -35,22 +35,42 @@
...
+    nextAvailableStart = TimeRange.START_OF_DAY; // Reset for finding optional available times
+    ImmutableList<TimeRange> optionalAvailableTimes = ImmutableList.copyOf(getAvailableTimeRanges(optionalUnavailableTimes, request));
+
+    
+    if (availableTimes.contains(TimeRange.WHOLE_DAY)){ 
+      // Return the optional time ranges only if there are no other time constraints
+      return optionalAvailableTimes;
+    }
+    else if (optionalAvailableTimes.isEmpty() || optionalAvailableTimes.contains(TimeRange.WHOLE_DAY)) {
```
Comment (christophervo-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: else if goes on the same line as the previous end brace

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -90,4 +109,45 @@ private static TimeRange findAvailableTimeRange(TimeRange currentUnavailableTime
...
+    // Get time ranges from events that include at least one attendee
+    ImmutableList<TimeRange> unavailableTimes = 
+        Streams.stream(events)
+            .filter(event -> !Collections.disjoint(attendees, event.getAttendees()))
+            .map(Event::getWhen)
+            .collect(toImmutableList());
+    return unavailableTimes;
+  }
+
+  private static ImmutableList<TimeRange> getOverlappingTimeRanges(ImmutableList<TimeRange> requiredTimes, ImmutableList<TimeRange> optionalTimes, MeetingRequest request) {
```
Comment (christophervo-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: is this longer than 80 chars? Should split/lint this

**Review by christophervo-g — APPROVED** [review]
LGTM

**Pull request merged**